### ISSUE TITLE
Support DrawerLayout Interactions

### ISSUE THREAD

**Issue by mateot1** [robolectric/robolectric#5102]
### Description
I have found no way to interact with a DrawerLayout with Robolectric. Espresso has actions specifically for them (see `DrawerActions`) and also will open a DrawerLayout when performing click actions on buttons that should open them as expected, but these do not work in Robolectric tests.

### Steps to Reproduce
Write a test with Espresso actions:
`onView(withId(R.id.drawer_layout_id)).perform(open())`
or 
`onView(withId(R.id.button_that_should_open_drawer_layout)).perform(click())`
then try to perform actions or verifications on DrawerLayout or subviews -- works in Espresso, not Robolectric

**Comment by daespark** [comment]
I agree. Haven't found a good way to test navigation drawer personally.

**Comment by anandwana001** [comment]
Any update on this? 
Did anyone have any work around for this or robolectric implemented some solution?

**Comment by mateot1** [comment]
maybe they're just waiting for everyone to switch to using bottom navigation...

**Comment by BenHenning** [comment]
Suggest looking at https://github.com/oppia/oppia-android/pull/2204. We came up with a workaround that might be somewhat generalizable: by calling computeScroll() directly on the DrawerLayout after opening/closing it should force the animation to complete in Robolectric (and conveniently in a way that doesn't break Espresso). Note that this is primarily for our use case (opening/closing the drawer) so I'm not sure if it helps with other DrawerLayout interactions.

We probably won't be using DrawerActions anymore for this purpose since it can't work with Robolectric (see the issue for specifics, but the tl;dr is that since Robolectric doesn't draw views, it doesn't complete the drawer's animation for opening).

**Comment by hoisie** [comment]
From oppia/oppia-android#2204

> Aha, it seems that computeScroll() is only called during the View's drawing step, and Robolectric skips view rendering. It's a bit bad that Android has a rendering process with side effects imo, but that's something that we need to live with for certain animations it seems.

Is there something Robolectric can do to invoke computeScroll? Where is it called from the framework?

**Comment by utzcoz** [comment]
> Is there something Robolectric can do to invoke computeScroll? Where is it called from the framework?

I found the invoking is here [View.draw(Canvas canvas, ViewGroup parent, long drawingTime) called by ViewGroup.drawChild](https://cs.android.com/android/platform/superproject/+/master:frameworks/base/core/java/android/view/View.java;l=22093;drc=master;bpv=0;bpt=1). This method will call `computeScroll()` when the canvas doesn't support hardware accelerated or attach info doesn't need hardware accelerated. Another important invoking is here [View.updateDisplayListIfDirty for hardware accelerated supported](https://cs.android.com/android/platform/superproject/+/master:frameworks/base/core/java/android/view/View.java;l=21208;drc=master).

**Comment by BenHenning** [comment]
@utzcoz's findings align with mine: the key problem (as I understand it) is that since Robolectric doesn't call View's draw methods, the side-effect computeScroll() never gets called. If Robolectric were willing to simulate drawing by calling the draw methods but not actually doing anything with the results, I suspect that would fix the issue (though that sounds like a very scary change that may broadly impact test performance).

There was one odd thing that I found too: Oppia Android uses LooperMode PAUSED for all of our tests, and calling computeScroll() is supposed to scheduled an animation runnable to continue scrolling until it completes (using the view choreographer's internal messaging queue). From previous investigations, Robolectric seems to run this queue correctly but for some reason in this circumstance it didn't properly trigger subsequent computeScroll() calls. I didn't dig into it further after I found out that a single call to computeScroll() resolved our immediate issue.

**Comment by hoisie** [comment]
I'm guessing doing drawing would be complex -- there would have to be something like a separate thread that observes whether the draw state is dirty for a View Hierarchy and then schedules drawing passees on the main looper via ViewRootImpl.

I think for now invoking the draw method (or computeScroll) is the most practical workaround. Maybe Robolectric could make that easier?

**Comment by hoisie** [comment]
For example, I wonder if something like this would work as a workaround as well:
```
Canvas c = new Canvas();
// Possibly set a Bitmap if thee canvas needs explicit bounds
drawerLayoutView.draw(c); // or the root activity view
```
